File: winRT_FastMines/src_winrt/ua/ksn/fmg/view/win_rt/draw/mosaics/CellPaint.cs
using System;
using Windows.UI.ViewManagement;
using Windows.UI.Xaml.Media.Imaging;
using ua.ksn.fmg.model.mosaics.cell;
using ua.ksn.fmg.model.mosaics;
using ua.ksn.geom;

namespace ua.ksn.fmg.view.win_rt.draw.mosaics
{

   /// <summary>
   /// Helper class for drawing info
   /// </summary>
   public class CellPaint
   {
      protected GraphicContext gContext;

      public CellPaint(GraphicContext gContext)
      {
         this.gContext = gContext;
         DefaultBackgroundFillColor = new UISettings().UIElementColor(UIElementType.ButtonFace).Cast();
      }

      private const string DRAW_FONT_NAME = "NirmalaUI";
      private const int DRAW_FONT_SIZE = 30;

      //static CellPaint()
      private bool _registerFont;
      private void RegisterFont()
      {
         if (_registerFont) return;
         try
         {
            BitmapFont.RegisterFont(DRAW_FONT_NAME, DRAW_FONT_SIZE);
            _registerFont = true;
         }
         catch (Exception ex)
         {
            System.Diagnostics.Debug.Assert(false, ex.Message);
         }
      }

      public void Paint(BaseCell cell, WriteableBitmap bmp)
      {
         // TODO ограничиваю рисование только границами своей фигуры
         //...

         // all paint
         PaintComponent(cell, bmp);
         PaintBorder(cell, bmp);
      }

      public void PaintBorder(BaseCell cell, WriteableBitmap bmp) {
         // TODO set pen width
         //... = gContext.PenBorder.Width;

         // draw lines
         PaintBorderLines(cell, bmp);

         // debug - визуально провер€ю верность вписанного квадрата (провер€ть при ширине пера около 21)
         //var rcInner = cell.getRcInner(gContext.PenBorder.Width);
         //bmp.DrawRectangle(rcInner.x, rcInner.y, rcInner.right(), rcInner.bottom(), (Windows.UI.Color)Color.MAGENTA);
      }

      /// <summary> draw border lines </summary>
      public void PaintBorderLines(BaseCell cell, WriteableBitmap bmp) {
         if (gContext.IconicMode) {
            bmp.DrawPolyline(RegionAsXyxyxySequence(gContext.Bound, cell.getRegion()), (Windows.UI.Color)(cell.State.Down ? gContext.PenBorder.ColorLight : gContext.PenBorder.ColorShadow));
         } else {
            var color = cell.State.Down ? gContext.PenBorder.ColorLight : gContext.PenBorder.ColorShadow;
            var s = cell.getShiftPointBorderIndex();
            var v = cell.Attr.getVertexNumber(cell.getDirection());
            for (var i=0; i < v; i++) {
               var p1 = cell.getRegion().getPoint(i);
               var p2 = (i != (v - 1)) ? cell.getRegion().getPoint(i + 1) : cell.getRegion().getPoint(0);
               if (i == s)
                  color = cell.State.Down ? gContext.PenBorder.ColorShadow : gContext.PenBorder.ColorLight;
               bmp.DrawLine(p1.x + gContext.Bound.width, p1.y + gContext.Bound.height, p2.x + gContext.Bound.width, p2.y + gContext.Bound.height, (Windows.UI.Color)color);
            }
         }
      }

      private void PaintComponent(BaseCell cell, WriteableBitmap bmp)
      {
         PaintComponentBackground(cell, bmp);

         var rcInner = cell.getRcInner(gContext.PenBorder.Width).moveXY(gContext.Bound);

         // output Pictures
         if ((gContext.ImgFlag != null) &&
             (cell.State.Status == EState._Close) &&
             (cell.State.Close == EClose._Flag))
         {
            var destRc = (Windows.Foundation.Rect) rcInner;
            var srcImg = gContext.ImgFlag;
            var srcRc = new Windows.Foundation.Rect(0, 0, srcImg.PixelWidth, srcImg.PixelHeight);
            bmp.Blit(destRc, srcImg, srcRc);
         }
         else if ((gContext.ImgMine != null) &&
                  (cell.State.Status == EState._Open) &&
                  (cell.State.Open == EOpen._Mine))
         {
            var destRc = (Windows.Foundation.Rect) rcInner;
            var srcImg = gContext.ImgMine;
            var srcRc = new Windows.Foundation.Rect(0, 0, srcImg.PixelWidth, srcImg.PixelHeight);
            bmp.Blit(destRc, srcImg, srcRc);
         }
         else
         // output text
         {
            string szCaption;
            Color txtColor;
            if (cell.State.Status == EState._Close)
            {
               txtColor = gContext.ColorText.getColorClose((int) cell.State.Close.Ordinal());
               szCaption = cell.State.Close.toCaption();
               //szCaption = cell.getCoord().x + ";" + cell.getCoord().y; // debug
               //szCaption = ""+cell.getDirection(); // debug
            }
            else
            {
               txtColor = gContext.ColorText.getColorOpen((int) cell.State.Open.Ordinal());
               szCaption = cell.State.Open.toCaption();
            }
            if (!string.IsNullOrEmpty(szCaption))
            {
               if (cell.State.Down)
                  rcInner.moveXY(1, 1);
               RegisterFont();
               bmp.DrawString(szCaption, rcInner.left(), rcInner.top(), DRAW_FONT_NAME, DRAW_FONT_SIZE,
                  (Windows.UI.Color) txtColor);
            }
         }
      }

      /// <summary> ÷вет заливки €чейки по-умолчанию. «ависит от текущего UI манагера </summary>
      private Color DefaultBackgroundFillColor { get; set; }

      /// <summary> залить €чейку нужным цветом </summary>
      protected void PaintComponentBackground(BaseCell cell, WriteableBitmap bmp)
      {
         //if (gContext.IconicMode) // когда русуетс€ иконка, а не игровое поле, - делаю попроще...
         //   return;
         var color = cell.getBackgroundFillColor(
            gContext.BkFill.Mode,
            DefaultBackgroundFillColor,
            gContext.BkFill.getColor
            );
         bmp.FillPolygon(RegionAsXyxyxySequence(gContext.Bound, cell.getRegion()), (Windows.UI.Color) color);
      }

      private static int[] RegionAsXyxyxySequence(Size bound, Region region)
      {
         var points = new int[region.CountPoints * 2 + 2];
         int i;
         for (i=0; i < region.CountPoints; i++) {
            var point = region.getPoint(i);
            points[i * 2] = point.x + bound.width;
            points[i * 2 + 1] = point.y + bound.height;
         }
         { // Add the first point also at the end of the array if the line should be closed.
            var point = region.getPoint(0);
            points[i * 2 + 0] = point.x + bound.width;
            points[i * 2 + 1] = point.y + bound.height;
         }
         return points;
      }
   }
}

File: swing_FastMines/src/swing/ua/ksn/fmg/view/swing/draw/mosaics/CellPaint.java
package ua.ksn.fmg.view.swing.draw.mosaics;

import java.awt.BasicStroke;
import java.awt.Color;
import java.awt.Font;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.Rectangle;
import java.awt.RenderingHints;
import java.awt.Shape;
import java.awt.Stroke;
import java.awt.font.FontRenderContext;
import java.awt.font.TextLayout;
import java.awt.geom.Rectangle2D;

import javax.swing.UIDefaults;
import javax.swing.UIManager;

import ua.ksn.fmg.model.mosaics.EClose;
import ua.ksn.fmg.model.mosaics.EOpen;
import ua.ksn.fmg.model.mosaics.EState;
import ua.ksn.fmg.model.mosaics.cell.BaseCell;
import ua.ksn.fmg.view.swing.draw.GraphicContext;
import ua.ksn.geom.Point;
import ua.ksn.geom.Rect;
import ua.ksn.swing.geom.Cast;

/**
 * Helper class for drawing info
 * @author SeregaLBN
 *
 */
public class CellPaint {
	protected GraphicContext gContext;
	
	public CellPaint(GraphicContext gContext) {
		this.gContext = gContext;
	}

	protected void repaint(BaseCell cell) {
//    	gContext.getOwner().paintImmediately(Cast.toRect(cell.getRcOuter()));
		gContext.getOwner().repaint(Cast.toRect(cell.getRcOuter()));
	}

	/** @see javax.swing.JComponent.paint */
	public void paint(BaseCell cell, Graphics g) {
//		Object obj = this;
//		if (obj instanceof JComponent) {
//			JComponent This = (JComponent)obj;
//			This.paint(g);
//		} else
		{
			Graphics2D g2d = (Graphics2D) g;

			// save
			Shape shapeOld = g2d.getClip();

			// ограничиваю рисование только границами своей фигуры
			g2d.setClip(Cast.toPolygon(cell.getRegion()));

			// all paint
			this.paintComponent(cell, g);
			this.paintBorder(cell, g);

			// restore
			g2d.setClip(shapeOld);
		}
	}

	/** @see javax.swing.JComponent.paintBorder */
	public void paintBorder(BaseCell cell, Graphics g) {
//		Object obj = this;
//		if (obj instanceof JComponent) {
//			JComponent This = (JComponent)obj;
//			This.paintBorder(g);
//			super.paintBorder(g);
//			return;
//		}

		Graphics2D g2d = (Graphics2D) g;
		// save
		Stroke strokeOld = g2d.getStroke();
		//RenderingHints rendHintsOld = g2d.getRenderingHints();
		Object renderingHintKeyAntialiasingOld = g2d.getRenderingHint(RenderingHints.KEY_ANTIALIASING);

		// set my custom params
		g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON); // TODO для релиза сменить на VALUE_ANTIALIAS_ON 
		g2d.setStroke(new BasicStroke(gContext.getPenBorder().getWidth())); // TODO глянуть расширенные параметры конструктора пера

		// draw lines
		paintBorderLines(cell, g);

		// restore
		g2d.setStroke(strokeOld);
//		g2d.setRenderingHints(rendHintsOld);
		g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, renderingHintKeyAntialiasingOld);

		// debug - визуально проверяю верность вписанного квадрата (проверять при ширине пера около 21)
//		Rect rcInner = cell.getRcInner(gContext.getPenBorder().getWidth());
//		g.setColor(Color.MAGENTA);
//		g.drawRect(rcInner.x, rcInner.y, rcInner.width, rcInner.height);
	}

	/** draw border lines */
	public void paintBorderLines(BaseCell cell, Graphics g) {
		if (gContext.isIconicMode()) {
			g.setColor(Cast.toColor(cell.getState().isDown() ? gContext.getPenBorder().getColorLight() : gContext.getPenBorder().getColorShadow()));
			g.drawPolygon(Cast.toPolygon(cell.getRegion()));
		} else {
			g.setColor(Cast.toColor(cell.getState().isDown() ? gContext.getPenBorder().getColorLight()  : gContext.getPenBorder().getColorShadow()));
			int s = cell.getShiftPointBorderIndex();
			int v = cell.getAttr().getVertexNumber(cell.getDirection());
			for (int i=0; i<v; i++) {
				Point p1 = cell.getRegion().getPoint(i);
				Point p2 = (i != (v-1)) ? cell.getRegion().getPoint(i+1) : cell.getRegion().getPoint(0);
				if (i==s)
					g.setColor(Cast.toColor(cell.getState().isDown() ? gContext.getPenBorder().getColorShadow(): gContext.getPenBorder().getColorLight()));
				g.drawLine(p1.x, p1.y, p2.x, p2.y);
			}
		}
	}

	/** @see javax.swing.JComponent.paintComponent */
	public void paintComponent(BaseCell cell, Graphics g) {
		Color colorOld = g.getColor();

		paintComponentBackground(cell, g);

		Rect rcInner = cell.getRcInner(gContext.getPenBorder().getWidth());
//		g.setColor(Color.MAGENTA);
//		g.drawRect(rcInner.x, rcInner.y, rcInner.width, rcInner.height);

		// output Pictures
		if ((gContext.getImgFlag() != null) &&
			(cell.getState().getStatus() == EState._Close) &&
			(cell.getState().getClose() == EClose._Flag))
		{
			gContext.getImgFlag().paintIcon(gContext.getOwner(), g, rcInner.x, rcInner.y);
		} else
		if ((gContext.getImgMine() != null) &&
			(cell.getState().getStatus() == EState._Open ) &&
			(cell.getState().getOpen() == EOpen._Mine))
		{
			gContext.getImgMine().paintIcon(gContext.getOwner(), g, rcInner.x, rcInner.y);
		} else
		// output text
		{
			String szCaption;
			if (cell.getState().getStatus() == EState._Close) {
				g.setColor(Cast.toColor(gContext.getColorText().getColorClose(cell.getState().getClose().ordinal())));
				szCaption = cell.getState().getClose().toCaption();
//				szCaption = cell.getCoord().x + ";" + cell.getCoord().y; // debug
//				szCaption = ""+cell.getDirection(); // debug
			} else {
				g.setColor(Cast.toColor(gContext.getColorText().getColorOpen(cell.getState().getOpen().ordinal())));
				szCaption = cell.getState().getOpen().toCaption();
			}
			if ((szCaption != null) && (szCaption.length() > 0))
			{
				if (cell.getState().isDown())
					rcInner.moveXY(1, 1);
				DrawText(g, szCaption, Cast.toRect(rcInner));
//				{ // test
//					Color clrOld = g.getColor(); // test
//					g.setColor(Color.red);
//					g.drawRect(rcInner.x, rcInner.y, rcInner.width, rcInner.height);
//					g.setColor(clrOld);
//				}
			}
		}

		// restore
		g.setColor(colorOld);
	}

	private ua.ksn.Color _defaultBkColor;
	/** Цвет заливки ячейки по-умолчанию. Зависит от текущего UI манагера */
	private ua.ksn.Color getDefaultBackgroundFillColor() {
		if (_defaultBkColor == null) {
			UIDefaults uiDef = UIManager.getDefaults();
	
			if (gContext.isIconicMode()) // когда русуется иконка, а не игровое поле, - делаю попроще...
				_defaultBkColor = Cast.toColor(uiDef.getColor("Panel.background"));
			else {
				String key = "Panel.background"; // "ToggleButton.light"; // "Button.light"; // 
				// ToggleButton.darkShadow : javax.swing.plaf.ColorUIResource[r=105,g=105,b=105]
				// ToggleButton.background : javax.swing.plaf.ColorUIResource[r=240,g=240,b=240]
				// ToggleButton.focus      : javax.swing.plaf.ColorUIResource[r=0,g=0,b=0]
				// ToggleButton.highlight  : javax.swing.plaf.ColorUIResource[r=255,g=255,b=255]
				// ToggleButton.light      : javax.swing.plaf.ColorUIResource[r=227,g=227,b=227]
				// ToggleButton.shadow     : javax.swing.plaf.ColorUIResource[r=160,g=160,b=160]
				// ToggleButton.foreground : javax.swing.plaf.ColorUIResource[r=0,g=0,b=0]
		
		//		if (state.getStatus() == EStatus._Close)
		//			if (state.isDown())
		//				key = "ToggleButton.light";
		//			else
		//				key = "ToggleButton.highlight";
		//		else
		//			key = "ToggleButton.shadow";
		
				Color clr = uiDef.getColor(key);
				if (clr == null) {
					System.out.println("Invalid color key: " + key);
					clr = uiDef.getColor("Panel.background");
				}
				_defaultBkColor = Cast.toColor(clr);
			}
		}
		return _defaultBkColor;
	}

	/** залить ячейку нужным цветом */
	protected void paintComponentBackground(BaseCell cell, Graphics g) {
		if (gContext.isIconicMode()) // когда русуется иконка, а не игровое поле, - делаю попроще...
			return;
		g.setColor(Cast.toColor(cell.getBackgroundFillColor(
				gContext.getBackgroundFill().getMode(),
				getDefaultBackgroundFillColor(),
				gContext.getBackgroundFill().getColors()
				)));
		g.fillPolygon(Cast.toPolygon(cell.getRegion()));
	}

	private static Rectangle2D getStringBounds(String text, Font font) {
		TextLayout tl = new TextLayout(text, font, new FontRenderContext(null, true, true));
		return tl.getBounds();
//		return font.getStringBounds(text, new FontRenderContext(null, true, true));
	}
	public static void DrawText(Graphics g, String text, Rectangle rc) {
		if ((text == null) || text.trim().isEmpty())
			return;
		//DrawText(m_GContext.m_hDCTmp, szCaption, -1, &sq_tmp, DT_CENTER | DT_VCENTER | DT_SINGLELINE);
		Rectangle2D bnd = getStringBounds(text, g.getFont());
//		{ // test
//			Color clrOld = g.getColor();
//			g.setColor(Color.BLUE);
//			g.fillRect(rc.x, rc.y, rc.width, rc.height);
//			g.setColor(clrOld);
//		}
		g.drawString(text,
				rc.x          +(int)((rc.width -bnd.getWidth ())/2.),
				rc.y+rc.height-(int)((rc.height-bnd.getHeight())/2.));
	}

}